### PR TITLE
Support struct-type to apply waterfall.end intrinsic

### PR DESCRIPTION
**Pull request by xuechen417**
The waterfall intrinsics can only accept simple types. The backend support has been merged. This is a front-end update. For a struct type, we should extract the v4f32 descriptor and i32 tfe members and generate waterfall.end instrinsic for each one.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit f83fea70592d3020ab6e913cd25354b377eca802
<details><summary>CTS tests (Failed: 0/208497)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 36978/69503 (53.2%)</li>
<li>Failed: 0/69503 (0.0%)</li>
<li>Not Supported: 32525/69503 (46.8%)</li><li>Warnings: 0/69503 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 41144/69504 (59.2%)</li>
<li>Failed: 0/69504 (0.0%)</li>
<li>Not Supported: 28360/69504 (40.8%)</li><li>Warnings: 0/69504 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 41178/69490 (59.3%)</li>
<li>Failed: 0/69490 (0.0%)</li>
<li>Not Supported: 28312/69490 (40.7%)</li><li>Warnings: 0/69490 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit a13f9bade063916b8bfdd8adfa0b8001c1c5564b
<details><summary>CTS tests (Failed: 0/208497)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 36978/69503 (53.2%)</li>
<li>Failed: 0/69503 (0.0%)</li>
<li>Not Supported: 32525/69503 (46.8%)</li><li>Warnings: 0/69503 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 41144/69504 (59.2%)</li>
<li>Failed: 0/69504 (0.0%)</li>
<li>Not Supported: 28360/69504 (40.8%)</li><li>Warnings: 0/69504 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 41178/69490 (59.3%)</li>
<li>Failed: 0/69490 (0.0%)</li>
<li>Not Supported: 28312/69490 (40.7%)</li><li>Warnings: 0/69490 (0.0%)</li>
</ul></details>


</ul></details>

**Review by nhaehnle — COMMENTED** [review]
Since `createWaterfallLoop` is expressed in fairly generic terms, could you please look into implementing the support for structs more generically as well?

I imagine implementing a recursive helper function like:
```c++
enum class MapToSimpleMode {
  Int32,
  SimpleVector,
};

// Rename MapToInt32Func -> MapToSimpleTypeFunc
Value *BuilderBase::CreateMapToSimpleType(MapToSimpleTypeFunc mapFunc, ArrayRef<Value *> mappedArgs, ArrayRef<Value *> passthroughArgs, MapToSimpleMode mode) {
  ...
}

Value *BuilderBase::CreateMapToSimpleType(MapToSimpleTypeFunc mapFunc, ArrayRef<Value *> mappedArgs, ArrayRef<Value *> passthroughArgs, MapToSimpleMode mode) {
  return CreateMapToInt32(mapFunc, mappedArgs, passthroughArgs, MapToSimpleMode::Int32);
}
```
For the implementation, just start with the current implementation of CreateMapToInt32, add struct handling, and add guards based on the `mode`.

The implementation of `createWaterfallLoop` then boils down to a single call of the new `CreateMapToSimpleType` method.

File: lgc/builder/BuilderImpl.cpp [hunk 1]
```diff
@@ -686,30 +686,65 @@ Instruction *BuilderImpl::createWaterfallLoop(Instruction *nonUniformInst, Array
     SetInsertPoint(nonUniformInst->getNextNode());
     SetCurrentDebugLocation(nonUniformInst->getDebugLoc());
 
-    Use *useOfNonUniformInst = nullptr;
-    Type *waterfallEndTy = resultValue->getType();
-    if (auto vecTy = dyn_cast<FixedVectorType>(waterfallEndTy)) {
+    // The second element in the array is to represent tfe relevant variables if they exist.
```
Comment (tsymalla-AMD) [inline lgc/builder/BuilderImpl.cpp:689]:
```suggestion
    // The second element in the array is to represent the relevant variables if they exist.
```

Comment (tsymalla) [reply lgc/builder/BuilderImpl.cpp:689]: Ah, I see, that was intended.

**Comment by xuechen417** [comment]
Thank you for the detailed suggestions. I will re-implement as your advice.

**Comment by xuechen417** [comment]
V1: Implement new `CreateMapToSimpleType`

**Comment by amdvlk-admin** [comment]
### Test summary for commit 99854684fbc8bcfe22022634f2284d22f400ba99
<details><summary>CTS tests (Failed: 0/208496)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 36978/69503 (53.2%)</li>
<li>Failed: 0/69503 (0.0%)</li>
<li>Not Supported: 32525/69503 (46.8%)</li><li>Warnings: 0/69503 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 41144/69503 (59.2%)</li>
<li>Failed: 0/69503 (0.0%)</li>
<li>Not Supported: 28359/69503 (40.8%)</li><li>Warnings: 0/69503 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 41178/69490 (59.3%)</li>
<li>Failed: 0/69490 (0.0%)</li>
<li>Not Supported: 28312/69490 (40.7%)</li><li>Warnings: 0/69490 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 7b5d2a0cb891abdc2ebf849481c7c7dc6999050a
<details><summary>CTS tests (Failed: 0/208497)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Rhel 8.9, Gfx10</summary><ul>

<li>Passed: 36979/69503 (53.2%)</li>
<li>Failed: 0/69503 (0.0%)</li>
<li>Not Supported: 32524/69503 (46.8%)</li><li>Warnings: 0/69503 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 41144/69504 (59.2%)</li>
<li>Failed: 0/69504 (0.0%)</li>
<li>Not Supported: 28360/69504 (40.8%)</li><li>Warnings: 0/69504 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 41178/69490 (59.3%)</li>
<li>Failed: 0/69490 (0.0%)</li>
<li>Not Supported: 28312/69490 (40.7%)</li><li>Warnings: 0/69490 (0.0%)</li>
</ul></details>


</ul></details>

**Review by nhaehnle — APPROVED** [review]

**Comment by Sisyph** [comment]
I think this commit made it impossible to build LLPC with a version of LLVM that doesn't include the graphics changes. Can you please add an LLVM_MAIN_REVISION  guard somewhere to re-enable that?

```
compiler/lgc/builder/BuilderBase.cpp:305:36: error: no member named 'amdgcn_waterfall_end' in namespace 'llvm::Intrinsic'
    CreateIntrinsic(Intrinsic::amdgcn_waterfall_end, waterfallEndTy, {waterfallBegin, resultValue}, nullptr);
```

**Comment by xuechen417** [comment]
> I think this commit made it impossible to build LLPC with a version of LLVM that doesn't include the graphics changes. Can you please add an LLVM_MAIN_REVISION guard somewhere to re-enable that?
> 
> ```
> compiler/lgc/builder/BuilderBase.cpp:305:36: error: no member named 'amdgcn_waterfall_end' in namespace 'llvm::Intrinsic'
>     CreateIntrinsic(Intrinsic::amdgcn_waterfall_end, waterfallEndTy, {waterfallBegin, resultValue}, nullptr);
> ```

Thanks @Sisyph for reporting the issue. @dstutt Could help tell me the specific llvm version that I should use to guard `amdgcn_waterfall_end`?

**Comment by Sisyph** [comment]
> > I think this commit made it impossible to build LLPC with a version of LLVM that doesn't include the graphics changes. Can you please add an LLVM_MAIN_REVISION guard somewhere to re-enable that?
> > ```
> > compiler/lgc/builder/BuilderBase.cpp:305:36: error: no member named 'amdgcn_waterfall_end' in namespace 'llvm::Intrinsic'
> >     CreateIntrinsic(Intrinsic::amdgcn_waterfall_end, waterfallEndTy, {waterfallBegin, resultValue}, nullptr);
> > ```
> 
> Thanks @Sisyph for reporting the issue. @dstutt Could help tell me the specific llvm version that I should use to guard `amdgcn_waterfall_end`?

You added some support in LLVM right? If so, I think it would be appropriate to go the the commit where you add that, then go to the next merge of llvm/main into amd-gfx, check llvm/include/llvm/Config/llvm-config.h.cmake for the value of LLVM_MAIN_REVISION, and use that number.

**Comment by xuechen417** [comment]
> You added some support in LLVM right? If so, I think it would be appropriate to go the the commit where you add that, then go to the next merge of llvm/main into amd-gfx, check llvm/include/llvm/Config/llvm-config.h.cmake for the value of LLVM_MAIN_REVISION, and use that number.

Thank for giving the guardline. Actually, @dstutt added the support in LLVM. I have no idea of the commit info.

**Comment by dstutt** [comment]
I think we normally put these changes (that require amd-gfx branch) behind LLVM_HAVE_BRANCH_AMD_GFX
Or do you need a specific version?

**Comment by xuechen417** [comment]
> > I think we normally put these changes (that require amd-gfx branch) behind LLVM_HAVE_BRANCH_AMD_GFX Or do you need a specific version?
> 
> I think it has been a reasonable length of time so there is not much need for the specific version. LLVM_HAVE_BRANCH_AMD_GFX sounds good to me

Good. `CreateWaterfallEnd` is already behind LLVM_HAVE_BRANCH_AMD_GFX in `createWaterfallLoop`. I don't think we need take any action.

**Pull request merged**